content/6_Advanced/Lagrange.mdx

```diff
@@ -27,6 +23,145 @@ adding lambda\*smth
+
+$f(x)$ will be the maximum cost $x$-flow through the graph. We can repeatidly find the maximum cost augmenting path $x$ times to get our answer. Because the maximum cost as a function of flow is concave, $f(x)$ will be concave. You can read more about simulating cost flows [here](https://codeforces.com/blog/entry/118391).
+
+</Spoiler>
+
+Consider the following graphs of $f(x)$ and $f(x)-\lambda x$. In this example, we have $\lambda=5$.
+
+<iframe
+	src="https://www.desmos.com/calculator/ydynwc2fej?embed"
+	width="500px"
+	height="300px"
+	frameborder="0"
+/>
+
+Here is where the fact that $f(x)$ is concave comes in. Because the slope is non-increasing, we know that $f(x) - \lambda x$ will first increase, then stay the same, and finally decrease.
+
+Let $v(\lambda)$ be the optimal maximal achievable sum with $\lambda$ penalty and $c(\lambda)$ be the number of subarrays used to achieve $v(\lambda)$ (note that if there are multiple such possibilities, we set $c$ to be the **minimal** number of subarrays to achieve $v$). These values can be calculated in $\mathcal{O}(N)$ time using the dynamic programming approach described above.
+
+When we assign the penalty of $\lambda$, we are trying to find the maximal sum if creating a subarray reduces our sum by $\lambda$. In other words, **we are trying to find the maximum of $f(x) - \lambda x$**.
+
+Without loss of generality, suppose there exists a slope equal to $\lambda$. Given the shape of $f(x) - \lambda x$, we know that $f(x) - \lambda x$ will be maximized at the points where $\lambda$ is equal to the slope of $f(x)$ (these points are red in the graph above). This means that $c(\lambda)$ will be the point at which $\lambda$ is equal to the slope of $f(x)$ (if there are multiple such points, then $c(\lambda)$ will be the leftmost one).
+
+Now we know exactly what $\lambda$ represents: $\lambda$ is the slope and $c(\lambda)$ is the position with slope equal to $\lambda$ (if there are multiple such positions then $c(\lambda)$ is the leftmost one).
+
+We binary search for $\lambda$ and find the highest $\lambda$ such that $c(\lambda) \le K$. Let the optimal value be $\lambda_{\texttt{opt}}$. Then our answer is $v(\lambda_{\texttt{opt}}) + \lambda_{\texttt{opt}} K$. Note that this works even if $c(\lambda_{\texttt{opt}}) \neq K$ since  $c(\lambda_{\texttt{opt}})$ and $K$ will be on the same line with slope $\lambda_{\texttt{opt}}$ in that case.
+
+Because calculating $v(\lambda)$ and $c(\lambda)$ with the dynamic programming solution described above will take $\mathcal{O}(N)$ time, this solution runs in $\mathcal{O}(N\log{\sum A[i]})$ time.
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+#define ll long long
+
+const int MAX = 3e5 + 5;
+
+int n, k, A[MAX];
+pair<ll, ll> dp[MAX][2];
+
+pair<ll, ll> better(pair<ll, ll> a, pair<ll, ll> b) {
+	if (a.first == b.first) return (a.second < b.second ? a : b);
+
+	return (a.first > b.first ? a : b);
+}
+
+bool solve(ll lmb) {
```

Comment:
move this into main

Reply:
Huh wdym? Like do you mean create a lambda in main or move it directly inside the binary search? It feels kind of weird and I edited my code so can you check if I still need to do this?

Reply:
move it directly inside the binsearch

while you're at it make `n`, `k`, `A`, and `dp` all local variables

Reply:
wait hollup u need to run the same dp again after the binary search tho to get ur answer so shud i do it as a lambda lol

Reply:
come again? i don't see why you can't just extract the code from the function into the for loop
nothing in it seems to depend on anything

Reply:
The function solveLambda is called in two separate places -- once inside the binary search and once in:
cout << solveLambda(lo).first + lo * k << endl;

So I will have to paste the same dp stuff into 2 separate places which is 😳

Reply:
you could also store the result in a variable `last_run` or whatever

Reply:
Hmm i guess u can do that but u would have to make sure that you only store the result from the last true run (assuming its smth like FFFFTTTT) and you have to make sure 'hi' is greater than the max opt lambda. I think it makes the code feel strange or smth. 

Can I just leave it as a lambda in main? tbh i think it looks the best and is most natural when it's as a function outside of main.

Reply:
alright then
but at least add some comments about what this function actually does, what it returns, etc.